### PR TITLE
Ao/seizure classification

### PR DESCRIPTION
**Pull request by ambroiseodt**
## Checklist:
- [x] Create Model.py: implementation of Model proposed in [https://arxiv.org/pdf/2106.11170.pdf](Transformer_base_model)
- [x] Update Train.py: create main function 
- [x] Update hyperparameter_tuning.py: create main function
- [x] Create parser function 
- [x] Create requirements.txt file
- [x] Add function to access gpu on ssh server
- [x] Fix issue with CUDA 11.1 and Pytorch for Train.py and hyperparameters_tuning.py (see https://github.com/ivadomed/MEEG-Brainstorm/issues/18)
- [x] In hyperparameters_tuning.py: fix issue with ray tune.run and usage of gpu (trials are not processed)
## Description
This pull request enables to create the model proposed in [https://arxiv.org/pdf/2106.11170.pdf](Transformer_base_model), to have training and evaluation functions as well as functions to optimize hyperparameters.
Functions will be callable from command line using a parser.
The requirements.txt file is useful to use this repository and the access to gpu device is enabled on ssh server.

### CODE REVIEW

**Comment by ambroiseodt** [comment]
@mpompolas I am currently having an error when using gpu cuda:0 on rosenberg. I move all the data onto the device and when I call the loss (torch.nn.CrossEntropyLoss()) there is the following issue when doing the backward:
Capture d’écran 2022-03-08 à 09 18 12
The code works perfectly when using a cpu (local and also on rosenberg) but as soon as I use a gpu, I have this error. on the contrary, the testing functions (such a the test_accuracy function in hyperparameter_tuning.py) works perfectly
Any idea ? Thanks

File: utils.py [hunk 1]
```diff
@@ -25,14 +29,90 @@ def get_class_distribution(data, label, display):
...
+
+def compute_l1_loss(w):
+    
+    """
+    Apply L1 regularization.
+    
+    Args: w: weight of the model.
+    """
+    
+    return torch.abs(w).sum()
```
Comment (mpompolas) [inline utils.py:114]: L1 loss is the sum of the absolutes of the difference between the predicted and the actual output.
I see here you are just take the sum of the absolute of the weights

File: utils.py [hunk 2]
```diff
@@ -25,14 +29,90 @@ def get_class_distribution(data, label, display):
...
+    
+    """
+    Apply L2 regularization.
+    
+    Args: w: weight of the model.
+    """
     
-    return count_label
\ No newline at end of file
+    return torch.square(w).sum()
```
Comment (mpompolas) [inline utils.py:124]: Same thing here:

L2 loss is the summation of the square of the difference between the predicted and the actual output.

Check the formula for the computation of the loss

File: utils.py (hunk 1 above)
Comment (ambroiseodt) [reply utils.py:114]: It does not compute the l1 loss (the name is bad chosen), but compute the L1 regularization component of the loss;
I can change the name of the function because it is badly titled


Capture d’écran 2022-03-15 à 16 32 13
s

File: utils.py (hunk 2 above)
Comment (ambroiseodt) [reply utils.py:124]: Same answer as above, I will change the name of the functions to avoid confusion

**Comment by mpompolas** [comment]
Hi @AmbroiseOdonnat.

I would suggest you add a pipeline figure. This helps users to follow the order that the data get processed, especially in a project like this where we go from EEG signals, to deep models, back to timing events. If you think it's still too early, add it once you have a full pipeline. Suggestion: On figure 1 of the paper you base you approach, add some numbers on the separate elements. After that, add a description of which functions contribute to which part of the figure.

File: Model.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+        # Compute Query dimension after average pooling
+        query_size = vector_size - kernel_size
+        query_size /= stride
+        query_size += 1
+        self.scaling = query_size ** (1 / 2) 
+        
+        # Weights initiation
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                nn.init.xavier_normal_(m.weight, gain = 1)
```
Comment (mpompolas) [inline Model.py:105]: Justification for xavier normal?

Comment (ambroiseodt) [reply Model.py:105]: It was the one in the original model and I tested all the other methods proposed by torch: it seems to be the best one

File: Model.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        
+        # Apply projection
+        x = self.projection(x)
+        return x
+    
+
+    
+""" ********** Multi-head attention ********** """
+
+class MultiHeadAttention(nn.Module):
```
Comment (mpompolas) [inline Model.py:202]: reference to paper that has the method:
"Attention is all you need"
https://arxiv.org/pdf/1706.03762.pdf

File: Model.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+""" ********** Feed-forward block ********** """
+
+class GELU(nn.Module):
+    
+    """
+    GeLU activation function.
+    """
+    
+    def forward(self, input: Tensor) -> Tensor:
+        return input*0.5*(1.0+torch.erf(input/math.sqrt(2.0)))
```
Comment (mpompolas) [inline Model.py]: I am not familiar with GELU.

A quick look at the paper describing it:
https://arxiv.org/pdf/1606.08415v3.pdf
shows a different formula.

Please confirm yours gets same/approximate values.

File: Model.py [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+""" ********** Temporal transforming ********** """
+
+class TransformerEncoderBlock(nn.Sequential):
+    
+    def __init__(self, slice_size, num_heads, dropout, forward_expansion, forward_dropout):
+
+        """    
+        Args:
+            slice_size (int): Size of slices,
+            num_heads (int): Number of heads in mulit-head block,
```
Comment (mpompolas) [inline Model.py:291]: typo

Comment (ambroiseodt) [reply Model.py:291]: Done

File: Model.py (hunk 4 above)
Comment (ambroiseodt) [reply Model.py:202]: Done

**Comment by ambroiseodt** [comment]
>  @mpompolas 
Thanks for the advice, the pipeline is not finished but I will do it as soon as possible (I want to think of a way to localize spikes)
Should I include this pipeline in the README.md ?

**Comment by mpompolas** [comment]
> > @mpompolas
> > Thanks for the advice, the pipeline is not finished but I will do it as soon as possible (I want to think of a way to localize spikes)
> > Should I include this pipeline in the README.md ?

Yes, it's the first thing people will see, so it would be good for them to get the overview straight from the beginning.

**Comment by mpompolas** [comment]
I want to run the code, but I have authorization trouble to pull the repo since it is in pull request mode.
I will merge it and test after.

OK?

**Comment by ambroiseodt** [comment]
@mpompolas  just wait a minute until I figure out the GELU function please

File: Model.py (hunk 5 above)
Comment (ambroiseodt) [reply Model.py]: This is what is seemed to me: in the model, nn.GELU() is used (GELU function coded in Pytorch) so the GELU function implemented is not used. I erased it as it is useless

**Comment by mpompolas** [comment]
Notifications that I get during fresh installarion:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests-oauthlib 1.3.0 requires oauthlib>=3.0.0, which is not installed.
tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
tensorflow-gpu 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
tensorboard 2.6.0 requires google-auth<2,>=1.6.3, but you have google-auth 2.6.2 which is incompatible.
Successfully installed aiohttp-3.7.0 aiohttp-cors-0.7.0 aioredis-1.3.1 async-timeout-3.0.1 blessings-1.7 chardet-3.0.4 charset-normalizer-2.0.12 click-8.0.4 colorama-0.4.4 colorful-0.5.4 deprecated-1.2.13 einops-0.4.1 filelock-3.6.0 google-api-core-2.7.1 google-auth-2.6.2 googleapis-common-protos-1.56.0 gpustat-0.6.0 hiredis-2.0.0 idna-3.3 matplotlib-3.5.1 mne-0.24.1 msgpack-1.0.3 multidict-6.0.2 numpy-1.21.2 nvidia-ml-py3-7.352.0 opencensus-0.8.0 opencensus-context-0.1.2 pandas-1.3.5 protobuf-3.19.4 psutil-5.9.0 py-spy-0.3.11 pydantic-1.9.0 pyyaml-6.0 ray-1.5.2 redis-4.1.4 requests-2.27.1 scikit-learn-1.0.2 scipy-1.6.2 seaborn-0.11.2 tabulate-0.8.9 tensorboardX-2.5 torch-1.8.1+cu102 torchvision-0.9.1+cu102 urllib3-1.26.9 yarl-1.7.2

```

**Comment by ambroiseodt** [comment]
> Notifications that I get during fresh installarion:
> 
> ```
> ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
> requests-oauthlib 1.3.0 requires oauthlib>=3.0.0, which is not installed.
> tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
> tensorflow-gpu 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
> tensorboard 2.6.0 requires google-auth<2,>=1.6.3, but you have google-auth 2.6.2 which is incompatible.
> Successfully installed aiohttp-3.7.0 aiohttp-cors-0.7.0 aioredis-1.3.1 async-timeout-3.0.1 blessings-1.7 chardet-3.0.4 charset-normalizer-2.0.12 click-8.0.4 colorama-0.4.4 colorful-0.5.4 deprecated-1.2.13 einops-0.4.1 filelock-3.6.0 google-api-core-2.7.1 google-auth-2.6.2 googleapis-common-protos-1.56.0 gpustat-0.6.0 hiredis-2.0.0 idna-3.3 matplotlib-3.5.1 mne-0.24.1 msgpack-1.0.3 multidict-6.0.2 numpy-1.21.2 nvidia-ml-py3-7.352.0 opencensus-0.8.0 opencensus-context-0.1.2 pandas-1.3.5 protobuf-3.19.4 psutil-5.9.0 py-spy-0.3.11 pydantic-1.9.0 pyyaml-6.0 ray-1.5.2 redis-4.1.4 requests-2.27.1 scikit-learn-1.0.2 scipy-1.6.2 seaborn-0.11.2 tabulate-0.8.9 tensorboardX-2.5 torch-1.8.1+cu102 torchvision-0.9.1+cu102 urllib3-1.26.9 yarl-1.7.2
> ```

@mpompolas Concerning the tensroflow issues, I think that this is because of your numpy version; I have to check mine and add into the requirements the convenient one (I forget to do so)
Same thing concerning oauthlib I have to check which version I have

**Comment by mpompolas** [comment]
> @mpompolas Concerning the tensroflow issues, I think that this is because of your numpy version; I have to check mine and add into the requirements the convenient one (I forget to do so)
> Same thing concerning oauthlib I have to check which version I have

Yes. I'm reporting it here so you can take care of it.

File: Train.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+        accuracy = correct / total 
+        F1_score = np.mean(f1_test)
+        
+        logger.info('Evaluation finished')
+        print("Accuracy on test set: {} \nF1 score on test set: {}".format(accuracy, F1_score))
+        
+        return accuracy , F1_score
+        
+
+def main():
```
Comment (mpompolas) [inline Train.py:379]: 1. Add all of the possible combinations of inputs as examples here, so people know how to call your pipeline.
2. ~~Define each input parameter (type, role)~~ Ignore (2) - I just saw you have this on your parser function

Comment (ambroiseodt) [reply Train.py:379]: For instance example of config file, save_path,.. ?

**Comment by ambroiseodt** [comment]
> Notifications that I get during fresh installarion:
> 
> ```
> ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
> requests-oauthlib 1.3.0 requires oauthlib>=3.0.0, which is not installed.
> tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
> tensorflow-gpu 2.3.1 requires numpy<1.19.0,>=1.16.0, but you have numpy 1.19.5 which is incompatible.
> tensorboard 2.6.0 requires google-auth<2,>=1.6.3, but you have google-auth 2.6.2 which is incompatible.
> Successfully installed aiohttp-3.7.0 aiohttp-cors-0.7.0 aioredis-1.3.1 async-timeout-3.0.1 blessings-1.7 chardet-3.0.4 charset-normalizer-2.0.12 click-8.0.4 colorama-0.4.4 colorful-0.5.4 deprecated-1.2.13 einops-0.4.1 filelock-3.6.0 google-api-core-2.7.1 google-auth-2.6.2 googleapis-common-protos-1.56.0 gpustat-0.6.0 hiredis-2.0.0 idna-3.3 matplotlib-3.5.1 mne-0.24.1 msgpack-1.0.3 multidict-6.0.2 numpy-1.21.2 nvidia-ml-py3-7.352.0 opencensus-0.8.0 opencensus-context-0.1.2 pandas-1.3.5 protobuf-3.19.4 psutil-5.9.0 py-spy-0.3.11 pydantic-1.9.0 pyyaml-6.0 ray-1.5.2 redis-4.1.4 requests-2.27.1 scikit-learn-1.0.2 scipy-1.6.2 seaborn-0.11.2 tabulate-0.8.9 tensorboardX-2.5 torch-1.8.1+cu102 torchvision-0.9.1+cu102 urllib3-1.26.9 yarl-1.7.2
> ```

This issue was caused because the virtual environment was created thanks to conda !
No error when creating a virtual environment with venv (command: `python3.X -m venv transformer_env`)
